src/VirtualContent.js

```diff
@@ -0,0 +1,341 @@
+          window.parseFloat(style.marginTop, 10) +
+          window.parseFloat(style.marginBottom, 10) +
+          childClientRect.height;
+        estimatedHeights.set(child, height);
+      }
+      return estimatedHeights.get(child);
+    };
+
+    const previouslyVisible = new Set();
+    for (let child = this.firstChild; child !== null; child = child.nextSibling) {
+      if (!child.hasAttribute('invisible')) {
+        previouslyVisible.add(child);
+      }
+    }
+
+    let beforePreviouslyVisible = previouslyVisible.size > 0;
+    let nextTop = 0;
+    let renderedHeight = 0;
+
+    // The estimated height of all elements made invisible since the last time
+    // an element was made visible (or start of the child list).
+    let currentInvisibleRunHeight = 0;
+    // The next empty space sentinel that should be reused, if any.
+    let nextEmptySpaceSentinel = this[_emptySpaceSentinelContainer].firstChild;
+    // Inserts an empty space sentinel representing the last contiguous run of
+    // invisible elements. Reuses already existing empty space sentinels, if
+    // possible.
+    const insertEmptySpaceSentinelIfNeeded = () => {
+      if (currentInvisibleRunHeight > 0) {
+        let sentinel = nextEmptySpaceSentinel;
+        if (nextEmptySpaceSentinel === null) {
+          sentinel = document.createElement('div');
+          this[_emptySpaceSentinelContainer].appendChild(sentinel);
+        }
+        nextEmptySpaceSentinel = sentinel.nextSibling;
+
+        const sentinelStyle = sentinel.style;
+        sentinelStyle.top = `${nextTop - currentInvisibleRunHeight}px`;
+        sentinelStyle.height = `${currentInvisibleRunHeight}px`,
+
+        this[_intersectionObserver].observe(sentinel);
+
+        currentInvisibleRunHeight = 0;
+      }
+    };
+
+    for (let child = this.firstChild; child !== null; child = child.nextSibling) {
+      if (beforePreviouslyVisible && previouslyVisible.has(child)) {
+        beforePreviouslyVisible = false;
+      }
+
+      let estimatedHeight = updateHeightEstimate(child);
+
+      const childClientTop = thisClientRect.top + nextTop;
+      const maybeInViewport =
+        (0 <= childClientTop + estimatedHeight) &&
+        (childClientTop <= window.innerHeight);
+      const childForceVisible = forceVisible.has(child);
+
+      if (maybeInViewport || childForceVisible) {
+        if (child.hasAttribute('invisible')) {
+          child.removeAttribute('invisible');
+          this[_resizeObserver].observe(child);
+          this[_intersectionObserver].observe(child);
+
+          const lastEstimatedHeight = estimatedHeight;
+          estimatedHeight = updateHeightEstimate(child);
+
+          if (beforePreviouslyVisible) {
+            const scrollingAncestor = nearestScrollingAncestor(this);
+            if (scrollingAncestor !== null) {
+              scrollingAncestor.scrollBy(0, estimatedHeight - lastEstimatedHeight);
+            }
+          }
+        }
+
+        const isInViewport =
```

Comment:
A comment should explain the difference between isInViewport and maybeInViewport, especially since they are computed in an identical way.

Reply:
When `maybeInViewport` is computed, `estimatedHeight` is either the current rendered height of the element or the last known (possibly inaccurate) height of the element, depending on whether or not the element was visible as of the last update. If `maybeInViewport` is true, then the element is made visible, its height is read, and the estimate* is updated again. The height estimate is now exact, and `isInViewport` is computed knowing the current, correct position of the element when rendered. If `isInViewport` is false, the element is made invisible and the true height of the element is added to the height of one of the empty space sentinel elements.

* Estimate is not a great choice of wording, there's no knowledgeable guessing involved.